### DISCUSSION TITLE
可變物件與不可變物件

### DISCUSSION THREAD

**Discussion by enoch-lightning** [1101-GS4719-C/Python-Community-GS4719-C#89] [General]
在**Python**中，所有東西都是**物件(object)**。物件有三個屬性：

1. identity=ID=物件在記憶體中的「位置=位址」
2. type=型別=物件如何使用記憶體(memory)
3.  #value=值=物件的值=物件本身
    a.值可變的物件稱為可變(mutable)物件。如
        i.list(串列)(Python原生的容器)：更強大與靈活的陣列
        ii.set(集合)(Python原生的容器)：用於判斷資料中是否含有某物件，且不考慮其出現順序
        iii.dict(dictionary=字典)(Python原生的容器)：一種以鍵(key)：值(value)對應的資料結構(在             pandas中很好用)
    b.值不可變的物件稱為不可變(immutable)物件。如
        i.tuple(元組)(Python原生的容器)：無法修改的list
        ii.string=str(字串)：因為字串的值不可變，必須先將字串的值複製一份到list中，用list處理，再轉回字串
    c.值可變的物件(可變物件)及值不可變的物件(不可變物件)，與記憶體的關係：
        i.改變可變物件的值，即直接更改物件的值。原變數名字仍「指向=參照=reference」原物件。不會在記憶體中另外建立新物件
        ii.改變不可變物件的值，不會更改原物件，而會建立新物件，並讓變數名字「指向=參照=reference」新物件

**Comment by enoch-lightning** [comment]
# 1
給定變數的名字，就是在記憶體中給定此變數名字「指向=參照=參考」」到的物件一個記憶體「位址=位置。」改變可變物件的值，會直接更改此物件的值，不會在記憶體中建立新物件。改變不可變物件的值，不是更改此物件的值，而是在記憶體中建立一新物件，再讓變數名字「指向=參照=參考」到此物件。
# 2
「指向=參照=參考」可以「貼標籤」的想法更生動地表示。物件的「型別=資料型態=data type=type」決定如何使用記憶體(是ram還是rom還是stack還是heap我搞不清楚)。若此型別的物件，其值「可變=可動」，代表可以直接修改原物件的值，不會另外建立新物件。變數名字的「標籤」仍然貼在原物件上。若此型別的物件，其值「不可變=不可動」，代表不可以直接修改原物件的值，必須另外在記憶體中建立此型別的新物件，並將變數名字的標籤從原物件上撕下，貼到剛建立的新物件上，也就是在記憶體的新地方放置新物件，讓變數名字「指向=參照=參考」到此，原物件也不會刪除，在重啟kernel時才會刪除。

**Comment by 107602505** [comment]
感謝重點整理

**Comment by Jie2222** [comment]
感謝同學整理，非常清楚~

**Comment by chupochun** [comment]
感謝同學的整理
